File: Arrays & Hashing/Python/TwoSum.py
from typing import List  # Importing List from typing for type annotations


class Solution:
    # Define a method that takes a list of integers and a target integer, and returns the indices of two numbers that add up to the target
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numlist = {}  # Initialize an empty dictionary to store the value and its corresponding index
        for index, value in enumerate(nums):  # Iterate over the list with both index and value
            # Check if the complement (target - current value) exists in the dictionary
            if target - value in numlist:
                # If found, return the indices: the index of the complement and the current index
                return [numlist[target - value], index]
            # Otherwise, store the current value and its index in the dictionary
            numlist[value] = index

# Test the solution locally
if __name__ == "__main__":
    nums = [2, 7, 11, 15]  # Example list of numbers
    target = 9  # The target sum we're looking for
    solution = Solution()  # Create an instance of the Solution class
    result = solution.twoSum(nums, target)  # Call the twoSum method with the list and target
    print(f"Indices of numbers that add up to {target}: {result}")  # Print the result (indices of two numbers)

File: Arrays & Hashing/Java/twoSum.java
import java.util.HashMap;

public class twoSum {
    public static int[] twooSum(int[] nums, int target) {
        // Create a HashMap to store the number and its index
        HashMap<Integer, Integer> m = new HashMap<>();
        
        // Iterate through the array
        for (int i = 0; i < nums.length; i++) {
            int v = nums[i];
            int x = target - v;
            
            // Check if the complement is already in the HashMap
            if (m.containsKey(x)) {
                return new int[] { m.get(x), i }; // Return the indices of the two numbers
            }
            // Add the current number and its index to the HashMap
            m.put(v, i);
        }
        // Return an empty array if no solution is found
        return new int[0];
    }
    
    public static void main(String[] args) {
        int[] nums = {2, 7, 11, 15};
        int target = 9;
        int[] result = twooSum(nums, target);
        
        if (result.length > 0) {
            System.out.println("Indices: " + result[0] + ", " + result[1]);
        } else {
            System.out.println("No solution found.");
        }
    }
}
